singer_sdk/streams/sql.py

```diff
@@ -501,26 +503,31 @@ def get_table_columns(self, full_table_name: str) -> dict[str, sqlalchemy.Column
         inspector = sqlalchemy.inspect(self._engine)
         columns = inspector.get_columns(table_name, schema_name)
 
-        result: dict[str, sqlalchemy.Column] = {}
-        for col_meta in columns:
-            result[col_meta["name"]] = sqlalchemy.Column(
+        return {
+            col_meta["name"]: sqlalchemy.Column(
                 col_meta["name"],
                 col_meta["type"],
                 nullable=col_meta.get("nullable", False),
             )
+            for col_meta in columns
+            if not column_names or col_meta["name"] in column_names
```

Comment:
I think it would be safer to have a case-insensitive comparison here.

- Snowflake is 'case unaware (at least by default), case sensitive'
- Redshift is 'case unaware, case insensitive'
- SQL Server is 'case aware, case insensitive'

We can certainly standardize on our side what casing we want to use while creating - but we can't control what happens on the server side or with a DBA.

For instance, even if the target prefers `lower_case` for SQL Server, a DBA could still manually create a target table with a column name `'User_Id'`. The operation to create a new `'user_id'` column in that case would be incorrect and would not be allowed on server side.

So, I think we want both of these: 

- targets can and should have a preference for how to case tables and columns.
- targets should not assume the preference is honored in the database itself.

What do you think?

Reply:
@aaronsteers this is still on the Tap side, called by the `SQLStream` class when composing the select statement during `get_records` [here](https://github.com/meltano/sdk/pull/1032/files#diff-61b20de8a12c7c91dfc4de62be5a230d2b4b1b2dd688628705554dc63a9490b6R1060-R1064). AFAIK we don’t do any conforming when reading from databases in Taps; the column names will always exactly match the stream property names because the singer schema is directly derived from the database schema.

I will start the work on column name conforming/transformations in the Target in the morning, where I will definitely run into comparison issues as you describe 😅

Reply:
https://docs.sqlalchemy.org/en/14/dialects/oracle.html#identifier-casing

> ## Identifier Casing
> In Oracle, the data dictionary represents all case insensitive identifier names using UPPERCASE text. SQLAlchemy on the other hand considers an all-lower case identifier name to be case insensitive. The Oracle dialect converts all case insensitive identifiers to and from those two formats during schema level communication, such as reflection of tables and indexes. Using an UPPERCASE name on the SQLAlchemy side indicates a case sensitive identifier, and SQLAlchemy will quote the name - this will cause mismatches against data dictionary data received from Oracle, so unless identifier names have been truly created as case sensitive (i.e. using quoted names), all lowercase names should be used on the SQLAlchemy side.

I think it's safe to compare the lower case identifiers?

Reply:
@edgarrmondragon - This is interesting... I wonder if we should just force everything to lower-case, since that represents probably the safest most reliable behavior across the board.

@kgpayne - Thanks for calling out the distinction between tap considerations (here) and target considerations (future). I still think forcing all comparisons and names to lower case may be safest default behavior across the board.